File: Languages/cpp/algorithms-dsa/leetcode/bit-manipulation/reverse_bits.cpp
/**
 * LeetCode 190: Reverse Bits
 * Topic: Bit Manipulation
 * Difficulty: Easy
 *
 * Reverse bits of a given 32-bit unsigned integer.
 * Bit-by-bit reversal: extract LSB and shift into result from MSB side.
 * Time: O(32) = O(1), Space: O(1)
 */
#include <iostream>
#include <cstdint>
#include <cassert>
using namespace std;

uint32_t reverseBits(uint32_t n) {
    uint32_t result = 0;
    for (int i = 0; i < 32; i++) {
        result = (result << 1) | (n & 1);
        n >>= 1;
    }
    return result;
}

int main() {
    assert(reverseBits(0b00000010100101000001111010011100) ==
                       0b00111001011110000010100101000000);

    assert(reverseBits(0b11111111111111111111111111111101) ==
                       0b10111111111111111111111111111111);

    assert(reverseBits(0) == 0);

    cout << "All tests passed!" << endl;
    return 0;
}

File: Languages/python/algorithms-dsa/leetcode/bit-manipulation/reverse_bits.py
"""
LeetCode #190 - Reverse Bits
Topic: Bit Manipulation
Difficulty: Easy

Reverse bits of a 32-bit unsigned integer bit by bit.

Time Complexity: O(1) - always 32 iterations
Space Complexity: O(1)
"""


def reverse_bits(n: int) -> int:
    result = 0
    for _ in range(32):
        result = (result << 1) | (n & 1)
        n >>= 1
    return result


if __name__ == "__main__":
    assert reverse_bits(0b00000010100101000001111010011100) == 964176192
    assert reverse_bits(0b11111111111111111111111111111101) == 3221225471
    assert reverse_bits(0) == 0
    print("All tests passed!")
